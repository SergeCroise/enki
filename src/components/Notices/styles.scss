.notices {
  .notices-row,
  .notices-editable-cell {
    width: 90%;
  }

  .notices-text-area {
    .eds-textarea {
      min-height: 1.25rem;
    }
  }

  .notices-icon-button {
    display: 'inline-block';
  }
}
