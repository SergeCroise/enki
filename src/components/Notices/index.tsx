import React, { useRef, useState } from 'react';
import { Tooltip } from '@entur/tooltip';
import { IconButton } from '@entur/button';
import { AddIcon, DeleteIcon } from '@entur/icons';
import {
  DataCell,
  EditableCell,
  Table,
  TableBody,
  TableRow,
} from '@entur/table';
import { TextArea } from '@entur/form';
import { Heading4 } from '@entur/typography';
import Notice from 'model/Notice';
import './styles.scss';
import { useSelector } from 'react-redux';
import { selectIntl } from 'i18n';
type Props = {
  notices?: Notice[];
  setNotices: (notices: Notice[]) => void;
};

export default ({ notices = [], setNotices }: Props) => {
  const [newNotice, setNewNotice] = useState<Notice>({ text: '' });
  const newNoticeInputEl = useRef<HTMLTextAreaElement>(null);
  const { formatMessage } = useSelector(selectIntl);

  const addNotice = () => {
    if (newNotice.text !== '') {
      setNotices([...notices, newNotice]);
      setNewNotice({ text: '' });
    }
  };

  const updateNotice = (index: number, text: string) => {
    if (text === '') {
      removeNotice(index);
      newNoticeInputEl?.current?.focus();
    } else {
      const copy = notices.slice();
      copy[index] = { text };
      setNotices(copy);
    }
  };

  const removeNotice = (index: number) => {
    const copy = notices.slice();
    copy.splice(index, 1);
    setNotices([...copy]);
  };

  return (
    <section className="notices">
      <Heading4>{formatMessage('noticesHeader')}</Heading4>
      <Table fixed>
        <TableBody>
          {notices?.map((notice, i) => (
            <TableRow key={'' + i} hover className="notices-row">
              <EditableCell className="notices-editable-cell">
                <TextArea
                  className="notices-text-area"
                  value={notice.text}
                  onChange={(e: any) => updateNotice(i, e.target.value)}
                />
              </EditableCell>
              <DataCell align="right">
                <Tooltip
                  placement="bottom"
                  content={formatMessage('deleteNoticeTooltip')}
                >
                  <IconButton
                    className="notices-icon-button"
                    onClick={() => removeNotice(i)}
                  >
                    <DeleteIcon />
                  </IconButton>
                </Tooltip>
              </DataCell>
            </TableRow>
          ))}
          <TableRow hover>
            <EditableCell className="notices-editable-cell">
              <TextArea
                className="notices-text-area"
                ref={newNoticeInputEl}
                label={formatMessage('newNoticeLabel')}
                value={newNotice.text}
                onChange={(e: any) => setNewNotice({ text: e.target.value })}
              />
            </EditableCell>
            <DataCell align="right">
              <Tooltip
                placement="bottom"
                content={formatMessage('addNoticeTooltip')}
              >
                <IconButton
                  className="notices-icon-button"
                  onClick={() => addNotice()}
                >
                  <AddIcon />
                </IconButton>
              </Tooltip>
            </DataCell>
          </TableRow>
        </TableBody>
      </Table>
    </section>
  );
};
