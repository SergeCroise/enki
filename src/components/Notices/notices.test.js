import { shallow } from 'enzyme';

import Notices from './';

const formatMessage = (id) => id;

describe('Notices', () => {
  it('should render no notices', () => {
    const wrapper = shallow(
      <Notices
        notices={[]}
        setNotices={() => {}}
        formatMessage={formatMessage}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should render single notices', () => {
    const wrapper = shallow(
      <Notices
        notices={[{ text: 'This a notice' }]}
        setNotices={() => {}}
        formatMessage={formatMessage}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render multiple notices', () => {
    const wrapper = shallow(
      <Notices
        notices={[
          { text: 'This a notice' },
          { text: 'This is another notice' },
        ]}
        setNotices={() => {}}
        formatMessage={formatMessage}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
