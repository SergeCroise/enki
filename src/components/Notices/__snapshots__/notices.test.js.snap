// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notices should render multiple notices 1`] = `
<section
  className="notices"
>
  <Heading4>
    noticesHeader
  </Heading4>
  <Table
    fixed={true}
  >
    <TableBody>
      <TableRow
        className="notices-row"
        hover={true}
        key="0"
      >
        <EditableCell
          className="notices-editable-cell"
        >
          <ForwardRef
            className="notices-text-area"
            onBlur={[Function]}
            onChange={[Function]}
            value="This a notice"
          />
        </EditableCell>
        <ForwardRef
          align="right"
        >
          <Tooltip
            content="deleteNoticeTooltip"
            placement="bottom"
          >
            <ForwardRef
              className="notices-icon-button"
              onClick={[Function]}
            >
              <DeleteIcon />
            </ForwardRef>
          </Tooltip>
        </ForwardRef>
      </TableRow>
      <TableRow
        className="notices-row"
        hover={true}
        key="1"
      >
        <EditableCell
          className="notices-editable-cell"
        >
          <ForwardRef
            className="notices-text-area"
            onBlur={[Function]}
            onChange={[Function]}
            value="This is another notice"
          />
        </EditableCell>
        <ForwardRef
          align="right"
        >
          <Tooltip
            content="deleteNoticeTooltip"
            placement="bottom"
          >
            <ForwardRef
              className="notices-icon-button"
              onClick={[Function]}
            >
              <DeleteIcon />
            </ForwardRef>
          </Tooltip>
        </ForwardRef>
      </TableRow>
      <TableRow
        hover={true}
        onClick={[Function]}
      >
        <ForwardRef />
        <ForwardRef
          align="right"
        >
          <Tooltip
            content="addNoticeTooltip"
            placement="bottom"
          >
            <ForwardRef
              className="notices-icon-button"
              onClick={[Function]}
            >
              <AddIcon />
            </ForwardRef>
          </Tooltip>
        </ForwardRef>
      </TableRow>
    </TableBody>
  </Table>
</section>
`;

exports[`Notices should render no notices 1`] = `
<section
  className="notices"
>
  <Heading4>
    noticesHeader
  </Heading4>
  <Table
    fixed={true}
  >
    <TableBody>
      <TableRow
        hover={true}
        onClick={[Function]}
      >
        <ForwardRef />
        <ForwardRef
          align="right"
        >
          <Tooltip
            content="addNoticeTooltip"
            placement="bottom"
          >
            <ForwardRef
              className="notices-icon-button"
              onClick={[Function]}
            >
              <AddIcon />
            </ForwardRef>
          </Tooltip>
        </ForwardRef>
      </TableRow>
    </TableBody>
  </Table>
</section>
`;

exports[`Notices should render single notices 1`] = `
<section
  className="notices"
>
  <Heading4>
    noticesHeader
  </Heading4>
  <Table
    fixed={true}
  >
    <TableBody>
      <TableRow
        className="notices-row"
        hover={true}
        key="0"
      >
        <EditableCell
          className="notices-editable-cell"
        >
          <ForwardRef
            className="notices-text-area"
            onBlur={[Function]}
            onChange={[Function]}
            value="This a notice"
          />
        </EditableCell>
        <ForwardRef
          align="right"
        >
          <Tooltip
            content="deleteNoticeTooltip"
            placement="bottom"
          >
            <ForwardRef
              className="notices-icon-button"
              onClick={[Function]}
            >
              <DeleteIcon />
            </ForwardRef>
          </Tooltip>
        </ForwardRef>
      </TableRow>
      <TableRow
        hover={true}
        onClick={[Function]}
      >
        <ForwardRef />
        <ForwardRef
          align="right"
        >
          <Tooltip
            content="addNoticeTooltip"
            placement="bottom"
          >
            <ForwardRef
              className="notices-icon-button"
              onClick={[Function]}
            >
              <AddIcon />
            </ForwardRef>
          </Tooltip>
        </ForwardRef>
      </TableRow>
    </TableBody>
  </Table>
</section>
`;
